File: backend/internal/model/restaurant.go
package model

// Restaurant represents the structure of a restaurant entity
// @Description This struct is used to represent a restaurant in the system
type Restaurant struct {
	// Unique identifier of the restaurant
	ID          string     `json:"id"`
	// Name of the restaurant
	Name        string  `json:"name"`
	// Latitude and longitude coordinates of the restaurant
	Latitude    float64 `json:"latitude"`
	// Longitude of the restaurant
	Longitude   float64 `json:"longitude"`
	// Address of the restaurant
	Address     string  `json:"address"`
	// Overall rating of the restaurant
	Rating      float64 `json:"rating"`
	// Number of reviews for the restaurant
	ReviewCount int     `json:"review_count"`
	// City where the restaurant is located
	CityID      string  `json:"city_id"`
	// District where the restaurant is located
	DistrictID  string  `json:"district_id"`
	// Food type name of the restaurant
	FoodType    string  `json:"food_type_name"`
	// Distance from user's location in kilometers
	Distance    float64 `json:"distance"`
}

type RestaurantDetail struct{
	// Information about the restaurant
	// This includes the restaurant's ID, name, address, etc.
	Restaurant Restaurant `json:"restaurant"`
	// Ratings for different aspects of the restaurant
	// Ambience, delivery, food, price, and service ratings
	Labels LabelsRating `json:"labels"`
	// List of platforms where the restaurant is available
	Platforms []string `json:"platforms"`
	// Ratings for the restaurant on different platforms
	// The length of this array should match the length of the Platforms array
	RatingPlatforms []float64 `json:"rating_platforms"`
}

type LabelRating struct {
	Rating float64 `json:"rating"`
	Count  int     `json:"count"`
}

type LabelsRating struct {
	Ambience LabelRating `json:"ambience"`
	Delivery LabelRating `json:"delivery"`
	Food     LabelRating `json:"food"`
	Price    LabelRating `json:"price"`
	Service  LabelRating `json:"service"`
}

File: frontend/src/types/restaurant.ts
export interface Restaurant {
  id: string;
  name: string;
  latitude: number;
  longitude: number;
  address: string;
  rating: number;
  reviewCount: number;
  cityId: string;
  districtId: string;
  foodTypeName: string;
  distance: number;
}

export interface RestaurantLabel {
  rating: number;
  count: number;
}

export interface RestaurantDetails {
  restaurant: Restaurant;
  labels: {
    ambience: RestaurantLabel;
    delivery: RestaurantLabel;
    food: RestaurantLabel;
    price: RestaurantLabel;
    service: RestaurantLabel;
  };
  platforms: string[] | null;
  ratingPlatforms: number[] | null;
}

export interface RestaurantListResponse {
  message: string;
  data: Restaurant[];
}

export interface RestaurantDetailResponse {
  message: string;
  data: RestaurantDetails;
}

export interface RestaurantListParams {
  lat?: number;
  lng?: number;
  foodtype?: string;
  limit?: number;
}

export interface RestaurantApiResponse {
  id: string;
  name: string;
  latitude: number;
  longitude: number;
  address: string;
  rating: number;
  review_count: number;
  city_id: string;
  district_id: string;
  food_type_name: string;
  distance: number;
}

export interface RestaurantDetailsApiResponse {
  restaurant: RestaurantApiResponse;
  labels: {
    ambience: RestaurantLabel;
    delivery: RestaurantLabel;
    food: RestaurantLabel;
    price: RestaurantLabel;
    service: RestaurantLabel;
  };
  platforms: string[] | null;
  rating_platforms: number[] | null;
}

export type RestaurantReviewLabel =
  | "ambience"
  | "delivery"
  | "food"
  | "price"
  | "service";

export interface RestaurantReview {
  feedback: string;
  label: string;
  rating: number;
  ratingId: string;
  ratingLabel: number;
  reviewTime: string;
  username: string;
}

export interface RestaurantReviewResponse {
  feedback: string;
  label: string;
  rating: number;
  rating_id: string;
  rating_label: number;
  review_time: string;
  username: string;
}

export interface RestaurantReviews {
  reviews: RestaurantReview[];
  totalReviews: number;
}

export interface RestaurantReviewsResponse {
  reviews: RestaurantReviewResponse[];
  total_reviews: number;
}

export interface RestaurantReviewsParams {
  id: string;
  label: RestaurantReviewLabel;
  page: number;
  count?: number;
}

export interface MenuItem {
  name: string;
  price: number;
}
